src/ctap/credential_id.rs

```diff
@@ -0,0 +1,254 @@
+            &credential_id[..],
+        );
+        credential_id.extend(&id_hmac);
+        Ok(credential_id)
+    }
+
+    /// Returns None if
+    /// - the format does not match any known versions, or
+    /// - the HMAC test fails.
+    ///
+    /// For v0 (legacy U2F) the credential ID consists of:
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted ECDSA private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v1 (ECDSA (algorithm -7)) the credential ID consists of:
+    /// -  1 byte : version number
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted ECDSA private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v2 (EdDSA over curve Ed25519 (algorithm -8, curve 6)) the credential ID consists of:
+    /// -  1 byte : version number
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted Ed25519 private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v3 (CBOR key handle) the credential ID consists of:
+    /// -  1 byte : version number
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 64 bytes * N: encrypted CBOR key handle
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    pub(super) fn decrypt_from_bytes(
+        env: &mut impl Env,
+        credential_id: Vec<u8>,
+    ) -> Result<Option<Self>, Ctap2StatusCode> {
+        if credential_id.len() < MIN_CREDENTIAL_ID_SIZE {
+            return Ok(None);
+        }
+        let hmac_message_size = credential_id.len() - 32;
+        if !verify_hmac_256::<Sha256>(
+            &env.key_store().key_handle_authentication()?,
+            &credential_id[..hmac_message_size],
+            array_ref![credential_id, hmac_message_size, 32],
+        ) {
+            return Ok(None);
+        }
+
+        if credential_id.len() == LEGACY_CREDENTIAL_ID_SIZE {
+            decrypt_legacy_key_handle(env, &credential_id[..hmac_message_size], ES256_ALGORITHM)
+        } else if is_legacy_key_handle_version(credential_id[0]) {
+            let algorithm = match credential_id[0] {
+                ECDSA_CREDENTIAL_ID_VERSION => ES256_ALGORITHM,
+                #[cfg(feature = "ed25519")]
+                ED25519_CREDENTIAL_ID_VERSION => EDDSA_ALGORITHM,
+                _ => return Ok(None),
+            };
+            decrypt_legacy_key_handle(env, &credential_id[1..hmac_message_size], algorithm)
+        } else if credential_id[0] == CBOR_CREDENTIAL_ID_VERSION {
+            decrypt_cbor_key_handle(env, &credential_id[1..hmac_message_size])
+        } else {
+            Ok(None)
+        }
```

Comment:
Looks to me that this whole diff could be rewritten more idiomatically as:

```rust
if len == LEGACY { return decrypt_legacy(); }
match credential_id[0] {
    ECDSA => decrypt_legacy(... ES256),
    ED25519 => decrypt_legacy(... EDDSA),
    CBOR => decrypt_cbor(),
    _ => Ok(None),
}
```

Reply:
Done